streams/src/test/java/org/apache/kafka/streams/integration/JoinGracePeriodDurabilityIntegrationTest.java

```diff
@@ -0,0 +1,275 @@
+    public String processingGuaranteee;
+
+    @Test
+    @SuppressWarnings("deprecation")
+    public void shouldRecoverBufferAfterShutdown() {
+        final String testId = safeUniqueTestName(getClass(), testName);
+        final String appId = "appId_" + testId;
+        final String streamInput = "Streaminput" + testId;
+        final String tableInput = "Tableinput" + testId;
+        final String storeName = "grace";
+        final String output = "output" + testId;
+
+        // create multiple partitions as a trap, in case the buffer doesn't properly set the
+        // partition on the records, but instead relies on the default key partitioner
+        cleanStateBeforeTest(CLUSTER, 2, streamInput, tableInput, output);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+        final KStream<String, String> stream = builder.stream(streamInput, Consumed.with(STRING_SERDE, STRING_SERDE));
+        final KTable<String, String> table = builder.table(tableInput, Consumed.with(STRING_SERDE, STRING_SERDE), Materialized.as(
+            Stores.persistentVersionedKeyValueStore(storeName, Duration.ofMinutes(5))));
+        final KStream<String, String> joinedStream = stream.join(table,
+            MockValueJoiner.TOSTRING_JOINER,
+            Joined.with(Serdes.String(), Serdes.String(), Serdes.String(), "Grace", Duration.ofMillis(5))
+        );
+
+        final AtomicInteger eventCount = new AtomicInteger(0);
+        joinedStream.foreach((key, value) -> eventCount.incrementAndGet());
+
+        joinedStream.to(output);
+
+        final Properties streamsConfig = mkObjectProperties(mkMap(
+            mkEntry(StreamsConfig.APPLICATION_ID_CONFIG, appId),
+            mkEntry(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers()),
+            mkEntry(StreamsConfig.POLL_MS_CONFIG, Long.toString(COMMIT_INTERVAL)),
+            mkEntry(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, processingGuaranteee),
+            mkEntry(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath()),
+            mkEntry(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.StringSerde.class),
+            mkEntry(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.StringSerde.class)
+        ));
+
+        streamsConfig.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, COMMIT_INTERVAL);
+
+        KafkaStreams driver = getStartedStreams(streamsConfig, builder, true);
+        try {
+            produceSynchronouslyToPartitionZero(
+                tableInput,
+                asList(
+                    new KeyValueTimestamp<>("k1", "v1", scaledTime(0L)),
+                    new KeyValueTimestamp<>("k2", "v2", scaledTime(0L)),
+                    new KeyValueTimestamp<>("k3", "v3", scaledTime(0L)),
+                    new KeyValueTimestamp<>("k4", "v4", scaledTime(0L)),
+                    new KeyValueTimestamp<>("k5", "v5", scaledTime(0L)),
+                    new KeyValueTimestamp<>("k6", "v6", scaledTime(0L))
```

Comment:
I could be missing something but shouldn't this test produce another table value with a later timestamp to prove the join results are table key `k2` time 0L and stream `k2` time 2 and not the latest table `k2` with a later timestamp?